File: typescript/tssearch/src/filetypes.ts
/*
 * filetypes.ts
 *
 * identifies file types (archive, binary, text, unknown)
 */

"use strict";

import * as config from './config';
import * as common from './common';
import {FileType} from './filetype';
import {FileUtil} from './fileutil';

interface FileTypeMap {
    [key: string]: string[]
}

export class FileTypes {
    private static fileTypeMap: FileTypeMap = FileTypes.getFileTypeMap();

    private static getFileTypeMap(): FileTypeMap {
        const fs = require('fs');

        let json = '';
        if (fs.existsSync(FileUtil.expandPath(config.FILETYPESJSONPATH))) {
            json = fs.readFileSync(FileUtil.expandPath(config.FILETYPESJSONPATH)).toString();
        } else {
            throw new Error('File not found: ' + config.FILETYPESJSONPATH);
        }

        const fileTypeMap: FileTypeMap = {};

        const obj = JSON.parse(json);
        if (obj.hasOwnProperty('filetypes') && Array.isArray(obj['filetypes'])) {
            obj['filetypes'].forEach(ft => {
                const typename: string = ft['type'];
                const extensions: string[] = ft['extensions'];
                fileTypeMap[typename] = common.setFromArray(extensions);
            });
        } else throw new Error("Invalid filetypes file: " + config.FILETYPESJSONPATH);

        fileTypeMap.text = fileTypeMap.text.concat(fileTypeMap.code, fileTypeMap.xml);
        fileTypeMap.searchable = fileTypeMap.text.concat(fileTypeMap.binary, fileTypeMap.archive);

        return fileTypeMap;
    }

    public static fromName(name: string): FileType {
        const uname: string = name.toUpperCase();
        if (uname === 'ARCHIVE')
            return FileType.Archive;
        if (uname === 'BINARY')
            return FileType.Binary;
        if (uname === 'CODE')
            return FileType.Code;
        if (uname === 'TEXT')
            return FileType.Text;
        if (uname === 'XML')
            return FileType.Xml;
        return FileType.Unknown;
    }

    public static toName(fileType: FileType): string {
        if (fileType === FileType.Archive)
            return 'ARCHIVE';
        if (fileType === FileType.Binary)
            return 'BINARY';
        if (fileType === FileType.Code)
            return 'CODE';
        if (fileType === FileType.Text)
            return 'TEXT';
        if (fileType === FileType.Xml)
            return 'XML';
        return 'UNKNOWN';
    }

    public static getFileType(filename: string): FileType {
        if (FileTypes.isCodeFile(filename))
            return FileType.Code;
        if (FileTypes.isXmlFile(filename))
            return FileType.Xml;
        if (FileTypes.isTextFile(filename))
            return FileType.Text;
        if (FileTypes.isBinaryFile(filename))
            return FileType.Binary;
        if (FileTypes.isArchiveFile(filename))
            return FileType.Archive;
        return FileType.Unknown;
    }

    public static getFileTypeAsync(filename: string, cb: (ft: FileType) => void): void {
        if (FileTypes.isCodeFile(filename))
            return cb(FileType.Code);
        if (FileTypes.isXmlFile(filename))
            return cb(FileType.Xml);
        if (FileTypes.isTextFile(filename))
            return cb(FileType.Text);
        if (FileTypes.isBinaryFile(filename))
            return cb(FileType.Binary);
        if (FileTypes.isArchiveFile(filename))
            return cb(FileType.Archive);
        cb(FileType.Unknown);
    }

    public static isArchiveFile(filename: string): boolean {
        const ext: string = FileUtil.getExtension(filename);
        return FileTypes.fileTypeMap['archive'].indexOf(ext) > -1;
    }

    public static isBinaryFile(filename: string): boolean {
        const ext: string = FileUtil.getExtension(filename);
        return FileTypes.fileTypeMap['binary'].indexOf(ext) > -1;
    }

    public static isCodeFile(filename: string): boolean {
        const ext: string = FileUtil.getExtension(filename);
        return FileTypes.fileTypeMap['code'].indexOf(ext) > -1;
    }

    public static isSearchableFile(filename: string): boolean {
        const ext: string = FileUtil.getExtension(filename);
        return FileTypes.fileTypeMap['searchable'].indexOf(ext) > -1;
    }

    public static isTextFile(filename: string): boolean {
        const ext: string = FileUtil.getExtension(filename);
        return FileTypes.fileTypeMap['text'].indexOf(ext) > -1;
    }

    public static isXmlFile(filename: string): boolean {
        const ext: string = FileUtil.getExtension(filename);
        return FileTypes.fileTypeMap['xml'].indexOf(ext) > -1;
    }

    public static isUnknownFile(filename: string): boolean {
        const ext: string = FileUtil.getExtension(filename);
        return FileTypes.fileTypeMap['searchable'].indexOf(ext) === -1;
    }
}

File: cpp/cppsearch/src/FileTypes.cpp
#include "boost/algorithm/string.hpp"
#include "rapidjson/document.h"
#include "rapidjson/filereadstream.h"
#include "common.h"
#include "config.h"
#include "FileTypes.h"
#include "FileUtil.h"

using namespace rapidjson;

namespace cppsearch {
    FileTypes::FileTypes() {
        m_archive_extensions = {};
        m_binary_extensions = {};
        m_text_extensions = {};
        load_file_types();
    }

    void FileTypes::load_file_types() {
        auto file_types_path = std::string(XSEARCHPATH);
        file_types_path.append("/shared/filetypes.json");

        if (!FileUtil::file_exists(file_types_path)) {
            std::string msg = "Filetypes file not found: ";
            msg.append(file_types_path);
            log_error(msg);
            // TODO: SearchException
            return;
        }

        FILE* fp = fopen(file_types_path.c_str(), "r");

        char readBuffer[65536];
        FileReadStream is(fp, readBuffer, sizeof(readBuffer));

        Document document;

        document.ParseStream(is);
        fclose(fp);

        assert(document.HasMember("filetypes"));
        const Value& file_types = document["filetypes"];
        assert(file_types.IsArray());
        for (SizeType i = 0; i < file_types.Size(); i++) {
            const Value::ConstObject &file_type = file_types[i].GetObject();
            assert(file_type.HasMember("type"));
            const Value &typeValue = file_type["type"];
            std::string type = typeValue.GetString();

            assert(file_type.HasMember("extensions"));
            const Value& extensions = file_type["extensions"];

            for (SizeType j = 0; j < extensions.Size(); j++) {
                if (type == "archive") {
                    m_archive_extensions.insert(extensions[j].GetString());
                } else if (type == "binary") {
                    m_binary_extensions.insert(extensions[j].GetString());
                } else if (type == "code") {
                    m_code_extensions.insert(extensions[j].GetString());
                } else if (type == "text") {
                    m_text_extensions.insert(extensions[j].GetString());
                } else if (type == "xml") {
                    m_xml_extensions.insert(extensions[j].GetString());
                }
            }
        }
    }

    FileType FileTypes::from_name(const std::string& name) {
        std::string uname = boost::to_upper_copy(name);
        if (uname == "TEXT") {
            return FileType::TEXT;
        }
        if (uname == "BINARY") {
            return FileType::BINARY;
        }
        if (uname == "CODE") {
            return FileType::CODE;
        }
        if (uname == "XML") {
            return FileType::XML;
        }
        if (uname == "ARCHIVE") {
            return FileType::ARCHIVE;
        }
        return FileType::UNKNOWN;
    }

    FileType FileTypes::get_file_type(const std::string& file_path) {
        if (is_code_file(file_path)) {
            return FileType::CODE;
        }
        if (is_xml_file(file_path)) {
            return FileType::XML;
        }
        if (is_text_file(file_path)) {
            return FileType::TEXT;
        }
        if (is_binary_file(file_path)) {
            return FileType::BINARY;
        }
        if (is_archive_file(file_path)) {
            return FileType::ARCHIVE;
        }
        return FileType::UNKNOWN;
    }

    bool FileTypes::found_ext(const std::set<std::string>* ext_set, const std::string& ext) {
        auto found = ext_set->find(ext);
        return found != ext_set->end();
    }

    bool FileTypes::is_archive_file(const std::string& file_path) {
        std::string ext = FileUtil::get_extension(file_path);
        return found_ext(&m_archive_extensions, ext);
    }

    bool FileTypes::is_binary_file(const std::string& file_path) {
        std::string ext = FileUtil::get_extension(file_path);
        return found_ext(&m_binary_extensions, ext);
    }

    bool FileTypes::is_code_file(const std::string& file_path) {
        std::string ext = FileUtil::get_extension(file_path);
        return found_ext(&m_code_extensions, ext);
    }

    bool FileTypes::is_searchable_file(const std::string& file_path) {
        std::string ext = FileUtil::get_extension(file_path);
        return found_ext(&m_text_extensions, ext)
               || found_ext(&m_code_extensions, ext)
               || found_ext(&m_xml_extensions, ext)
               || found_ext(&m_binary_extensions, ext)
               || found_ext(&m_archive_extensions, ext);
    }

    bool FileTypes::is_text_file(const std::string& file_path) {
        std::string ext = FileUtil::get_extension(file_path);
        return found_ext(&m_text_extensions, ext)
               || found_ext(&m_code_extensions, ext)
               || found_ext(&m_xml_extensions, ext);
    }

    bool FileTypes::is_unknown_file(const std::string& file_path) {
        return get_file_type(file_path) == FileType::UNKNOWN;
    }

    bool FileTypes::is_xml_file(const std::string& file_path) {
        std::string ext = FileUtil::get_extension(file_path);
        return found_ext(&m_xml_extensions, ext);
    }
}
